stl/src/syncstream.cpp

```diff
@@ -0,0 +1,26 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// initialize syncstream mutex map
+
+#include <syncstream>
+
+#ifdef __cpp_lib_syncbuf
+#pragma warning(disable : 4074)
+#pragma init_seg(compiler)
+
+_STD_BEGIN
+// OBJECT DECLARATIONS
+static map<void*, _Mutex_count_pair> _Mutex_map{};

```

Comment:
If we're going to have long-lived allocations in separately compiled code, we probably need to mark them as "CRT blocks" to prevent them from showing up as memory leaks (as viewed by the CRT leak tracking machinery). There's an existing example of this in STL code.

Reply:
Here's an example:
https://github.com/microsoft/STL/blob/ac4fde764ae716512bdd67e0e9577bd53947bc3d/stl/src/cond.cpp#L39
https://github.com/microsoft/STL/blob/ac4fde764ae716512bdd67e0e9577bd53947bc3d/stl/src/cond.cpp#L52

Reply:
Thanks 😄, I was going to ask for an example.

Reply:
I have some concerns regarding this issue.

- We allocate `map<void*, _Mutex_count_pair>` as `_CRT_BLOCK`, but `map` allocates more elements as it grows. Would not this be detected as a memory leak?
- To ensure allocation/deallocation of these two variables, I assume we have to create an empty static struct which I named `_Init_global_mutex_map` on my local device. This struct will use 1 byte and will live for the entire program lifetime. Will this cause the memory leak false-positive as well? 😕

Reply:
Yes, the allocations for the map nodes are the ones that need to be marked as CRT blocks. This can be done by giving the map a custom allocator.

The map itself doesn't need to be dynamically allocated; it can have static storage duration. No need to have an empty static struct, if I understand what's happening correctly.